e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +223,80 @@ func (pf *PortForwarder) start() error {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	// create an inner context for the watch so we can cancel threads between iterations
+	// the subprocess will be killed inside portForwardToDeployment
+	watchCtx, watchCancel := context.WithCancel(pf.ctx)
+	defer watchCancel()
+	// restart the port-forward if:
+	// - the subprocess is killed
+	// - the pod is restarted
+	restartChan := pf.signalOnFailure(
+		func() error {
+			return pf.cmd.Wait()
+		},
+		func() error {
+			return pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+				testwatcher.WatchTimeout(0), testwatcher.WatchContext(watchCtx))
+		},
+	)
+	select {
+	case <-pf.ctx.Done():
+		return fmt.Errorf("exiting - PortForwarder for pod %s/%s received done signal", pf.ns, pf.pod)
+	case <-restartChan:
+		pf.logger.Infof("PortForwarder for pod %s/%s received restart signal. Restarting port-forward", pf.ns, pf.pod)
+	}
+	if err := pf.portForwardToDeployment(true); err != nil {
+		// this is a potentially fatal error since we failed to restart the port
+		// forward. we log the error and return in case the test is already done
+		// with the port forward. the test will otherwise error the next time it
+		// tries to use the local port.
+		return err
+	}
+	return nil
+}
+
+// signalOnFailure returns a chan which will be closed if any of the provided
+// functions return. The functions should block and only return if a failure
+// has occurred and the port-forward must be restarted
+func (pf *PortForwarder) signalOnFailure(fns ...func() error) chan struct{} {
+	restartChan := make(chan struct{})
+	// signalled indicates whether a signal has been sent
+	var signalled bool
+	var mux sync.Mutex
+	for _, fn := range fns {
+		fn := fn
+		go func() {
+			err := fn()
+			mux.Lock()
+			defer func() {
+				signalled = true
+				mux.Unlock()
+			}()
+			if signalled {
+				return
+			}
+			select {
+			case <-pf.ctx.Done():
+				// if the parent context closes, we no longer need to restart
```

Comment:
Looks like the restartChan is never closed if the context is done.
There's probably a clearer way to do this.